smaug/python/ops/control_flow_ops.py

```diff
@@ -30,9 +28,100 @@ def merge(input_tensors, name="merge"):
   Returns:
     A tensor that the available input tensor forwards to.
   """
-  return add_node(
-      name=name,
-      op=Merge,
-      input_tensors=input_tensors,
+  return common.add_node(
+      name=name, op=types_pb2.Merge, input_tensors=input_tensors,
       output_tensors_dims=[input_tensors[0].shape.dims],
       output_tensor_layout=input_tensors[0].shape.layout)[0]
+
+class ControlFlowContext:
+  def __init__(self, pred, branch):
+    """Construct a control flow context.
+
+    Args:
+      pred: Predication tensor.
+      branch: Value of 0 or 1 to denote the False/True branch of the control
+        flow.
+    """
+    self._pred = pred
+    self._branch = branch
+    # Names of all nodes that are within this context.
+    self._internal_nodes = set()
+    # Mappings of the name of a known external tensors to its switch node ouput.
+    self._external_tensors = {}
+    # The next outer context to this context.
+    self._outer_context = None
+
+  def __enter__(self):
+    """Enter this control flow context."""
+    self._outer_context = global_vars.get_graph().control_flow_context
+    global_vars.get_graph().control_flow_context = self
+    return self
+
+  def __exit__(self, *args):
+    """Exit this control flow context."""
+    global_vars.get_graph().control_flow_context = self._outer_context
+
+  def add_result(self, result):
+    """Make the result nodes available in the outer context.
+
+    The results of the inner context will be used in the outer context. To
+    prevent the outer context from thinking that these results are external
+    and consequently adding unnecessary switch nodes, this makes the inner
+    nodes that produce the results internal to outer context.
+    """
+    if self._outer_context:
+      for res in result:
+        self._outer_context.add_node_name(res.source[0].name)
+
+  def add_node_name(self, name):
+    self._internal_nodes.add(name)
+
+  def add_node(self, node, input_tensors):
+    """Add this node to the control flow context.
+
+    This makes the node known to the context. We create a switch node for any
+    of its input tensor if it's deemed external to the context.
+    """
+    # Don't add the node if it's already added. Don't add a data op to avoid
+    # deadlock.
```

Comment:
"deadlock" means two competing threads are stuck trying to acquire a resource owned by the other thread. Do you mean an infinite loop?

Reply:
Right. Sorry I meant infinite loop, not a deadlock.